src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,63 @@
 package com.es.phoneshop.model.product;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static ArrayListProductDao instance;
+    private volatile ArrayList<Product> products = new ArrayList<>();
+
+    private ArrayListProductDao() {
+    }
+
+    public static ArrayListProductDao getInstance() {
+        if (instance == null) {
+            synchronized (ArrayListProductDao.class) {
+                if (instance == null) {
+                    instance = new ArrayListProductDao();
+                }
+            }
+        }
+
+        return instance;
+    }
+
+    public synchronized void setProducts(ArrayList<Product> products) {
+        this.products = products;
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public synchronized Product getProduct(Long id) throws ProductNotFoundException {
+        Optional<Product> tmpProduct = products.stream()
+                .filter(product -> product.getId().equals(id))
+                .findFirst();
+
+        if (!tmpProduct.isPresent()) {
+            throw new ProductNotFoundException("Product not found");
+        }
+
+        return tmpProduct.get();
     }
 
     @Override
-    public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+    public synchronized List<Product> findProducts() {
+        return products.stream()
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public synchronized void save(Product product) {
+        if (product != null && products.stream().noneMatch(product1 -> product1.getId().equals(product.getId()))) {
```

Comment:
Move the stream chain out of if clause

Reply:
Also, check for null product id and notify the calling logic if the product is not saved successfully (e.g. by throwing an exception)